### PR TITLE
Implement mhc_fused_post_pre with SYCL

### PR DESCRIPTION
**Pull request by cyxlily**
## Motivation

Add a SYCL implementation for fused mHC post+pre processing on XPU.

The main goal is to improve the small-batch XPU path for mHC by introducing a fused post/pre flow, reducing overhead on token-light workloads, and aligning the implementation with the CUDA-side FMA-style path(https://github.com/sgl-project/sglang/blob/main/python/sglang/kernels/ops/layernorm/mhc.py#L1070).

## Modifications

- Add the SYCL/XPU `mhc_fused_post_pre` path and `MHCFusedPostPreFmaKernel` for the small-token regime.
- Register the op `mhc_fused_post_pre` and expose it through sgl_kernel. 
- Add unit tests: tests/test_mhc_fused_post_pre.py
- Add benchmark: benchmark/bench_mhc_fused_post_pre.py

## Accuracy

Correctness was validated by comparing the fused path against the unfused reference path:
- reference: `hc_post + mhc_pre`
- target: `mhc_fused_post_pre`

All unit tests passed.

Command: pytest tests/test_mhc_fused_post_pre.py
Results: 
========================================================================== test session starts ==========================================================================
platform linux -- Python 3.12.0, pytest-9.1.1, pluggy-1.6.0
rootdir: /home/lily/PR/mhc/sgl-kernel-xpu
configfile: pyproject.toml
plugins: anyio-4.14.0
collected 24 items

tests/test_mhc_fused_post_pre.py ........................                                                                                                         [100%]

========================================================================== 24 passed in 8.77s ===========================================================================

## Performance

The fused path improves latency for many small-token configurations over unfused reference path.

MHC FUSED_POST_PRE: FUSED VS NON-FUSED
|   batch_size |   seq_len |   T |   hidden_size | dtype    | with_norm   |   fused_ms |   non_fused_ms |   speedup_time |
|-------------:|----------:|----:|--------------:|:---------|:------------|-----------:|---------------:|---------------:|
|            1 |         1 |   1 |          4096 | bfloat16 | False       |   0.029432 |       0.044512 |        1.51237 |
|            1 |         1 |   1 |          4096 | bfloat16 | True        |   0.024492 |       0.048256 |        1.97028 |
|            1 |         1 |   1 |          7168 | bfloat16 | False       |   0.045708 |       0.06526  |        1.42776 |
|            1 |         1 |   1 |          7168 | bfloat16 | True        |   0.052    |       0.071448 |        1.374   |
|            1 |        16 |  16 |          4096 | bfloat16 | False       |   0.032396 |       0.047684 |        1.47191 |
|            1 |        16 |  16 |          4096 | bfloat16 | True        |   0.036348 |       0.051142 |        1.40701 |
|            1 |        16 |  16 |          7168 | bfloat16 | False       |   0.052572 |       0.070928 |        1.34916 |
|            1 |        16 |  16 |          7168 | bfloat16 | True        |   0.057824 |       0.076076 |        1.31565 |
|            8 |         1 |   8 |          4096 | bfloat16 | False       |   0.028756 |       0.045916 |        1.59675 |
|            8 |         1 |   8 |          4096 | bfloat16 | True        |   0.032344 |       0.04966  |        1.53537 |
|            8 |         1 |   8 |          7168 | bfloat16 | False       |   0.045656 |       0.067756 |        1.48405 |
|            8 |         1 |   8 |          7168 | bfloat16 | True        |   0.051168 |       0.07358  |        1.43801 |
|           32 |         1 |  32 |          4096 | bfloat16 | False       |   0.042432 |       0.050804 |        1.1973  |
|           32 |         1 |  32 |          4096 | bfloat16 | True        |   0.047008 |       0.05538  |        1.1781  |
|           32 |         1 |  32 |          7168 | bfloat16 | False       |   0.071344 |       0.076232 |        1.06851 |
|           32 |         1 |  32 |          7168 | bfloat16 | True        |   0.078988 |       0.083746 |        1.06024 |

### CODE REVIEW

**Comment by mkumargarg** [comment]
Hi @cyxlily are you doing same here as in PR https://github.com/sgl-project/sgl-kernel-xpu/pull/278 which is merged and available on 'main'? could you please cross check if it is not duplication. apart from that please also note that https://github.com/sgl-project/sgl-kernel-xpu/pull/199 s doing the "MHC Pre Big Fuse Kernel" and https://github.com/sgl-project/sgl-kernel-xpu/pull/186 "HC Split Sinkhorn Kernel". Thanks.
cc: @krishnanand-k

**Comment by cyxlily** [comment]
> Hi @cyxlily are you doing same here as in PR #278 which is merged and available on 'main'? could you please cross check if it is not duplication. apart from that please also note that #199 s doing the "MHC Pre Big Fuse Kernel" and #186 "HC Split Sinkhorn Kernel". Thanks. cc: @krishnanand-k

Thanks for the reminder, I found your submission. Let me take a deeper look at the MHC part to see if there are any areas where performance can be optimized. I'll update my PR.

File: python/sgl_kernel/mhc.py [hunk 1]
```diff
@@ -2,6 +2,20 @@
 
 import torch
 
+XPU_FUSED_FALLBACK_HIDDEN_SIZE = 7168
+XPU_FUSED_FALLBACK_TOKENS = 32
+XPU_SMALL_BATCH_TOKENS = 32
```
Comment (mingfeima) [inline python/sgl_kernel/mhc.py]: these are constants on CUDA?

File: src/sycl/MHCFusedPostPre.cpp [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+void mhc_fused_post_pre_fma_kernel(
+    const at::Tensor& x,
+    const at::Tensor& residual,
+    const at::Tensor& post_layer_mix_2d,
+    const at::Tensor& comb_res_mix_3d,
+    const at::Tensor& fn,
+    at::Tensor& residual_out,
+    at::Tensor& mixes_partial_out,
+    at::Tensor& sqrsum_partial_out,
+    int64_t split_k) {
```
Comment (mingfeima) [inline src/sycl/MHCFusedPostPre.cpp]: `mhc_fused_post_pre_fma_kernel` seems not useful, just move it to `mhc_fused_post_pre_fma`?

**Review by mingfeima — CHANGES_REQUESTED** [review]
`tests/run_suite.py`: test_mhc_fused_post_pre.py is not in the per-commit suite. CI runs python3 run_suite.py --suite per-commit, so this UT never executes. Please add TestFile("test_mhc_fused_post_pre.py") next to the other MHC tests.

File: python/sgl_kernel/mhc.py [hunk 3]
```diff
@@ -2,6 +2,20 @@
...
+
+
+def _choose_small_batch_splits(num_tokens: int, n_splits_hint: int) -> int:
+    if n_splits_hint > 0:
+        return n_splits_hint
+    if num_tokens <= 4:
+        return 32
+    if num_tokens <= 16:
+        return 8
+    return 4
```
Comment (mingfeima) [inline python/sgl_kernel/mhc.py]: does this need to be aligned with cuda: https://github.com/sgl-project/sglang/blob/main/python/sglang/kernels/ops/layernorm/mhc.py#L1395-L1398?

**Review by mingfeima — COMMENTED** [review]
please take a look at comment for `_choose_small_batch_splits`

File: python/sgl_kernel/mhc.py (hunk 3 above)
Comment (cyxlily) [reply python/sgl_kernel/mhc.py]: Thanks for the comments. Updated the _choose_small_batch_splits to be aligned with CUDA.

File: src/sycl/MHCFusedPostPre.cpp (hunk 2 above)
Comment (cyxlily) [reply src/sycl/MHCFusedPostPre.cpp]: Thanks for the comments. Removed mhc_fused_post_pre_fma_kernel.

File: python/sgl_kernel/mhc.py (hunk 1 above)
Comment (cyxlily) [reply python/sgl_kernel/mhc.py]: Thanks for the comments. Updated constants to be aligned with CUDA.

**Review by mingfeima — APPROVED** [review]
